src/agent.rs

```diff
@@ -41,6 +42,7 @@ pub struct Agent {
     pub(crate) headers: Vec<Header>,
     /// Reused agent state for repeated requests from this agent.
     pub(crate) state: Arc<Mutex<AgentState>>,
+    pub(crate) resolver: ArcResolver,
```

Comment:
We have a few different options for this type:

 1. Resolver
 2. Arc<Resolver>
 3. Arc<dyn Resolver>
 4. ArcResolver (similar to (3))

Would you mind describing the considerations that went into choosing (4)? My default would be (1) but I'm guessing there's something I'm missing.

Reply:
Looking at this again, I see that Unit and Request store `Arc<Mutex<AgentState>>` rather than an an `Arc<Agent>`. I think switching Unit and Request to store an `Arc<Agent>` instead would make for a nicer change here, and would probably remove the need for an Arc on the Resolver object. That could be a separate change or part of this one if you prefer.

Reply:
FYI I took a stab at this refactoring and realized it's a bit more complicated than I thought, since the `Arc` has to own the `Agent`, but we expect the user code to own the Agent.

I'll try another round with moving the Mutex inside the AgentState.

Reply:
>  1.  Resolver
>  2. Arc<Resolver>
>  3. Arc<dyn Resolver>
>  4. ArcResolver (similar to (3))
>
> Would you mind describing the considerations that went into choosing (4)?

@jsha This is how we need to do it in Rust.

1. Since `Resolver` is a trait, the size of the concrete implementation is not known. The only way to handle that unknown would be to make `Agent` generic over resolver. I.e. we would have something like `Agent<R> where R: Resolver`. That in turn would force all enclosing types to also be generic over `R` and it would "bubble" to the API surface making for an awkward API. 

2. This syntax is pre rust edition 2018. It works, but is deprecated and causes a warning. The idea is that the `dyn` keyword makes it clear that we're talking about a trait object.

3. This is the good way. It doesn't make the API generic over the trait, and it does so by creating a vtable and doing dynamic dispatch. It's heap allocated, so the size of the actual concrete type is "hidden" via pointer indirection.

4. This is the [new type idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). It relies on `Deref` to access the held data. I suspect @rawler used it because it also gives us a chance to implement `Debug` for all the wrapped concrete impls.

@rawler My preference would be 3, cause the `ArcResolver` creates a level of indirection which I don't think helps the reader of the code much. What do you think?

Reply:
4 were, as you said, chosen to be able to provide a blanket `Debug`, but mainly to provide a `Default` implementation. Without it, we need to manually implement `Default` for the entire `Request` structure.

I don't mind explicit implementation, would also solve the problem of the `Default` now being an unusable value. I just got the impression that leaning on derived implementation was idiomatic in this codebase.